File: Data Structures & Algorithms/three-integer-sum/submission-5.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        res = []
        nums.sort()

        a = 0

        while a < len(nums)-2:
            b, c = a+1, len(nums)-1
            while b < c:
                cur_sum = nums[a] + nums[b] + nums[c]
                if cur_sum == 0:
                    res.append([nums[a], nums[b], nums[c]])
                    b+=1
                    c-=1
                    while b < c and nums[b] == nums[b-1]:
                        b+=1
                if cur_sum < 0:
                    b+=1
                if cur_sum > 0:
                    c-=1

            a += 1
            while a < len(nums)-2 and nums[a] == nums[a-1]:
                a += 1

        return res

        

File: Data Structures & Algorithms/products-of-array-discluding-self/submission-5.js
class Solution {
    /**
     * @param {number[]} nums
     * @return {number[]}
     */
    
    productExceptSelf(arr) {
    const prefix = new Array(arr.length).fill(1);
    const suffix = new Array(arr.length).fill(1);

    for (let i = 1; i < arr.length; i++) {
        prefix[i] = arr[i-1] * prefix[i-1];
    }

    for (let j = arr.length-2; j >= 0; j--) {
        suffix[j] = arr[j+1] * suffix[j+1]
    }

    const res = []

    for (let i = 0; i < arr.length; i++) {
        res.push(prefix[i] * suffix[i])
    }

    return res

    }
}
